src/js/components/JSONEditor.js

```diff
@@ -0,0 +1,487 @@
+import AceEditor from 'react-ace';
+import React from 'react';
+import deepEqual from 'deep-equal';
+
+import {omit} from '../utils/Util';
+import JSONUtil from '../utils/JSONUtil';
+import JSONEditorUtil from '../utils/JSONEditorUtil';
+
+const METHODS_TO_BIND = [
+  'handleBlur',
+  'handleChange',
+  'handleEditorLoad',
+  'handleWindowResize',
+  'handleFocus'
+];
+
+/**
+ * How long to wait before clearing the `isTyping` flag, used internally to
+ * prohibit external updates to the editor while the user is typing.
+ *
+ * @const {number}
+ */
+const ISTYPING_TIMEOUT = 2000;
+
+/**
+ * This is a high-order component on top of `AceEditor` that abstracts the JSON
+ * string manipulation and allows object-level access to the underlaying data.
+ *
+ * It's features are:
+ *
+ * - Solve some bugs and peculiarities of AceEditor component
+ * - Maintain visual coherence between object updates
+ * - Callback on per-property updates as the user types
+ * - Receive errors through properties and show them in the gutter
+ *
+ * @example <caption>How to use the JSONEditor</caption>
+ *
+ * handleChange(newObject) {
+ *   console.log('New object:', newObject);
+ * }
+ *
+ * handlePropertyChange(path, value) {
+ *   console.log('Property /'+path.join('/'), 'changed to', value);
+ * }
+ *
+ * render() {
+ *   let value = {
+ *     some: {value: 'object'}
+ *   };
+ *   let errors = [
+ *     {path: ['some'], message: 'An error on object `some`'}
+ *   ];
+ *
+ *   return (
+ *     <JSONEditor
+ *       value={value}
+ *       errors={errors}
+ *       onChange={this.handleChange}
+ *       onPropertyChange={this.handlePropertyChange}
+ *       />
+ *   );
+ * };
+ *
+ */
+class JSONEditor extends React.Component {
+  constructor() {
+    super(...arguments);
+
+    // We are using the react-way of updating the component **only** when we
+    // need to define a new text to work upon (ex. when the owner component has
+    // updated the value).
+    //
+    // Updating the AceEditor on every render cycle seems to cause some trouble
+    // to it's internals, that I couldn't pinpoint yet.
+    this.state = {
+      initialText: ''
+    };
+
+    //
+    // The following properties are part of the `internal`, non-react state
+    // and is synchronized with the react through `componentWillReceiveProps`
+    //
+    this.externalErrors = [];
```

Comment:
Ok that's fair. This is a pattern we have in the past tried to avoid so I will add this to the team discussion agenda so we all converge on a pattern, whatever that be.